gio-subclass/src/application.rs

```diff
@@ -0,0 +1,599 @@
+        ApplicationClassExt::override_vfuncs(klass, token);
+    }
+
+    object_type_fns!();
+}
+
+unsafe extern "C" fn application_startup<T: ApplicationBase>(ptr: *mut gio_ffi::GApplication)
+where
+    T::ImplType: ApplicationImpl<T>,
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let application = &*(ptr as *mut T::InstanceStructType);
+    let wrap: T = from_glib_borrow(ptr as *mut T::InstanceStructType);
+    let imp = application.get_impl();
+
+    imp.startup(&wrap)
+}
+
+unsafe extern "C" fn application_activate<T: ApplicationBase>(ptr: *mut gio_ffi::GApplication)
+where
+    T::ImplType: ApplicationImpl<T>,
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let application = &*(ptr as *mut T::InstanceStructType);
+    let wrap: T = from_glib_borrow(ptr as *mut T::InstanceStructType);
+    let imp = application.get_impl();
+
+    imp.activate(&wrap)
+}
+
+unsafe extern "C" fn application_open<T: ApplicationBase>(
+    ptr: *mut gio_ffi::GApplication,
+    files: *mut *mut gio_ffi::GFile,
+    num_files: libc::c_int,
+    hint: *const libc::c_char,
+) where
+    T::ImplType: ApplicationImpl<T>,
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let application = &*(ptr as *mut T::InstanceStructType);
+    let wrap: T = from_glib_borrow(ptr as *mut T::InstanceStructType);
+    let imp = application.get_impl();
+
+    imp.open(
+        &wrap,
+        &FromGlibContainer::from_glib_none_num(files, num_files as usize),
+        &from_glib_none(hint),
+    )
+}
+
+unsafe extern "C" fn application_command_line<T: ApplicationBase>(
+    ptr: *mut gio_ffi::GApplication,
+    cmd_line: *mut gio_ffi::GApplicationCommandLine,
+) -> libc::c_int
+where
+    T::ImplType: ApplicationImpl<T>,
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let application = &*(ptr as *mut T::InstanceStructType);
+    let wrap: T = from_glib_borrow(ptr as *mut T::InstanceStructType);
+    let imp = application.get_impl();
+
+    imp.command_line(&wrap, &from_glib_borrow(cmd_line))
+}
+
+unsafe extern "C" fn application_local_command_line<T: ApplicationBase>(
+    ptr: *mut gio_ffi::GApplication,
+    arguments: *mut *mut *mut libc::c_char,
+    exit_status: *mut libc::c_int,
+) -> glib_ffi::gboolean
+where
+    T::ImplType: ApplicationImpl<T>,
+{
+    callback_guard!();
+    floating_reference_guard!(ptr);
+    let application = &*(ptr as *mut T::InstanceStructType);
+    let wrap: T = from_glib_borrow(ptr as *mut T::InstanceStructType);
+    let imp = application.get_impl();
+
+    //TODO: if arguments is changed in the handler function, how do we release the memory
+    // previously associated with arguments?
+    let mut args = FromGlibPtrContainer::from_glib_none(ptr::read(arguments));
+
+    match imp.local_command_line(&wrap, &mut args) {
+        Some(ret) => {
+            *exit_status = ret;
+            ptr::write(arguments, args.to_glib_full()); //take full ownership of the data
```

Comment:
As quoted above, this is actually not correct. It's only allowed to remove items from that array, not change it completely. So you would need to check here which ones got removed and update the content of the array accordingly by removing and shifting the following ones. All a bit ugly.

The easiest solution here would probably be to not use a `Vec<String>` here but instead a custom type around `Vec<String>` that derefs to `&[String]` and in addition only provides a function to remove items (as no other operations are allowed).